evp.go

```diff
@@ -243,118 +223,91 @@ func setupEVP(withKey withKeyFunc, padding C.int,
+	return crypt(ctx, out, in)
 }
 
-func verifyEVP(withKey withKeyFunc, padding C.int,
-	h hash.Hash, label []byte, saltLen C.int, ch crypto.Hash,
+func verifyEVP(withKey withKeyFunc, padding int,
+	h hash.Hash, label []byte, saltLen int, ch crypto.Hash,
 	init initFunc, verify verifyFunc,
 	sig, in []byte) error {
 
 	ctx, err := setupEVP(withKey, padding, h, nil, label, saltLen, ch, init)
 	if err != nil {
 		return err
 	}
-	defer C.go_openssl_EVP_PKEY_CTX_free(ctx)
-	return verify(ctx, base(sig), C.size_t(len(sig)), base(in), C.size_t(len(in)))
+	defer ctx.free()
+	return verify(ctx, sig, in)
 }
 
-func evpEncrypt(withKey withKeyFunc, padding C.int, h, mgfHash hash.Hash, label, msg []byte) ([]byte, error) {
-	encryptInit := func(ctx C.GO_EVP_PKEY_CTX_PTR) error {
-		if ret := C.go_openssl_EVP_PKEY_encrypt_init(ctx); ret != 1 {
-			return newOpenSSLError("EVP_PKEY_encrypt_init failed")
-		}
-		return nil
+func evpEncrypt(withKey withKeyFunc, padding int, h, mgfHash hash.Hash, label, msg []byte) ([]byte, error) {
+	encryptInit := func(ctx evpPkeyCtx) error {
+		return ctx.encryptInit()
 	}
-	encrypt := func(ctx C.GO_EVP_PKEY_CTX_PTR, out *C.uchar, outLen *C.size_t, in *C.uchar, inLen C.size_t) error {
-		if ret := C.go_openssl_EVP_PKEY_encrypt(ctx, out, outLen, in, inLen); ret != 1 {
-			return newOpenSSLError("EVP_PKEY_encrypt failed")
-		}
-		return nil
+	encrypt := func(ctx evpPkeyCtx, out, in []byte) ([]byte, error) {
+		return ctx.encrypt(out, in)
 	}
 	return cryptEVP(withKey, padding, h, mgfHash, label, 0, 0, encryptInit, encrypt, msg)
 }
 
-func evpDecrypt(withKey withKeyFunc, padding C.int, h, mgfHash hash.Hash, label, msg []byte) ([]byte, error) {
-	decryptInit := func(ctx C.GO_EVP_PKEY_CTX_PTR) error {
-		if ret := C.go_openssl_EVP_PKEY_decrypt_init(ctx); ret != 1 {
-			return newOpenSSLError("EVP_PKEY_decrypt_init failed")
-		}
-		return nil
+func evpDecrypt(withKey withKeyFunc, padding int, h, mgfHash hash.Hash, label, msg []byte) ([]byte, error) {
+	decryptInit := func(ctx evpPkeyCtx) error {
+		return ctx.decryptInit()
 	}
-	decrypt := func(ctx C.GO_EVP_PKEY_CTX_PTR, out *C.uchar, outLen *C.size_t, in *C.uchar, inLen C.size_t) error {
-		if ret := C.go_openssl_EVP_PKEY_decrypt(ctx, out, outLen, in, inLen); ret != 1 {
-			return newOpenSSLError("EVP_PKEY_decrypt failed")
-		}
-		return nil
+	decrypt := func(ctx evpPkeyCtx, out, in []byte) ([]byte, error) {
+		return ctx.decrypt(out, in)
 	}
 	return cryptEVP(withKey, padding, h, mgfHash, label, 0, 0, decryptInit, decrypt, msg)
 }
 
-func evpSign(withKey withKeyFunc, padding C.int, saltLen C.int, h crypto.Hash, hashed []byte) ([]byte, error) {
-	signtInit := func(ctx C.GO_EVP_PKEY_CTX_PTR) error {
-		if ret := C.go_openssl_EVP_PKEY_sign_init(ctx); ret != 1 {
-			return newOpenSSLError("EVP_PKEY_sign_init failed")
-		}
-		return nil
+func evpSign(withKey withKeyFunc, padding int, saltLen int, h crypto.Hash, hashed []byte) ([]byte, error) {
+	signtInit := func(ctx evpPkeyCtx) error {
```

Comment:
This was already here, but:

```suggestion
	signInit := func(ctx evpPkeyCtx) error {
```